File: MultilanguageDijkstraAlgorithm/TestCase.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MultilanguageDijkstraAlgorithm
{
    class TestCase
    {
        private string filename;

        public TestCase(string filename)
        {
            this.filename = filename;
        }

        internal void Run()
        {
            Console.Write("Running " + filename + "... ");
            string[] lines = System.IO.File.ReadAllLines(filename);
            string metadata = lines[0];
            // #start#end#27#start:a:c:d:end#
            Assignment assignment = new Assignment();
            foreach(string line in lines.Skip(1).Where(str => str != ""))
            {
                string[] split = line.Split(':');
                string node1s = split[0];
                int distance = int.Parse(split[1]);
                string node2s = split[2];
                assignment.AddEdge(node1s, distance, node2s);
            }
            string[] metadataSplit = metadata.Split(new char[] { '#' }, StringSplitOptions.RemoveEmptyEntries);
            assignment.StartNode = assignment.AllNodes[metadataSplit[0]];
            assignment.FinishNode = assignment.AllNodes[metadataSplit[1]];
            bool expectSuccess = metadataSplit[2] != "FAIL";

            DijkstraResult result = Dijkstra.Compute(assignment);
            
            if (result.Successful)
            {
                if (!expectSuccess)
                {
                    Console.WriteLine("fail! (unexpectedly succeeded)");
                    return;
                }
                if (result.PathLength != int.Parse(metadataSplit[2]))
                {
                    Console.WriteLine("fail! (bad path length)");
                    return;
                }
                string[] nodeNames = metadataSplit[3].Split(':');
                if (nodeNames.Length != result.BestPath.Count)
                {
                    Console.WriteLine("fail! (bad node count)");
                    return;
                }
                for (int i =0; i < nodeNames.Length; i++)
                {
                    if (nodeNames[i] != result.BestPath[i].Name)
                    {
                        Console.WriteLine("fail! (bad path)");
                        return;
                    }
                }
                Console.WriteLine("OK!");
            }
            else
            {
                if (expectSuccess)
                {
                    Console.WriteLine("fail! (unexpectedly failed)");
                }
                else
                {
                    Console.WriteLine("OK!");
                }
            }
        }
    }
}

File: VBNet/TestCase.vb
Public Class TestCase
    Private filename As String

    Public Sub New(filename As String)
        Me.filename = filename
    End Sub

    Friend Sub Run()
        Console.Write("[VB.NET] Running test case " + filename + "... ")
        Dim lines() As String = System.IO.File.ReadAllLines(filename)
        Dim firstLine = lines(0)
        ' #start#end#27#start:a:c:d:end# 
        Dim assignment As Assignment = New Assignment()
        For Each line In lines.Skip(1).Where(Function(text) text <> "")
            Dim split = line.Split(":")
            Dim s1 = split(0)
            Dim dist = Integer.Parse(split(1))
            Dim s2 = split(2)
            assignment.AddEdge(s1, dist, s2)
        Next

        Dim metadata As String() = firstLine.Split(New Char() {"#"}, StringSplitOptions.RemoveEmptyEntries)
        assignment.StartNode = assignment.AllNodes(metadata(0))
        assignment.FinishNode = assignment.AllNodes(metadata(1))
        Dim expectSuccess = metadata(2) <> "FAIL"
        Dim result = Dijkstra.Compute(assignment)

        If result.Successful Then
            If Not expectSuccess Then
                Console.WriteLine("fail! (unexpectedly succeeded)")
                Return
            End If
            If result.PathLength <> Integer.Parse(metadata(2)) Then
                Console.WriteLine("fail! (bad path length)")
                Return
            End If
            Dim nodeNames = metadata(3).Split(":")
            If nodeNames.Length <> result.BestPath.Count Then
                Console.WriteLine("fail! (bad node count)")
                Return
            End If
            For i = 0 To nodeNames.Length - 1
                If nodeNames(i) <> result.BestPath(i).Name Then
                    Console.WriteLine("fail! (bad path)")
                    Return
                End If
            Next
            Console.WriteLine("OK!")
        Else
            If expectSuccess Then
                Console.WriteLine("fail! (unexpectedly failed)")
            Else
                Console.WriteLine("OK!")
            End If
        End If
    End Sub
End Class
